\input{../settings/boilerplate}
\chapter{Treebase: An R package for discovery, access and manipulation of online phylogenies}
\section{Introduction}

Applications that use phylogenetic information as part of their analyses
are becoming increasingly central to both evolutionary and ecological
research. The exponential growth in genetic sequence data available for
all forms of life has driven rapid advances in the methods that can
infer the phylogenetic relationships and divergence times across
different taxa (Huelsenbeck and Ronquist 2001; Stamatakis 2006; Drummond
and Rambaut 2007). Once again the product of one field has become the
raw data of the next. Unfortunately, while the discipline of
bioinformatics has emerged to help harness and curate the wealth of
genetic data with cutting edge computer science, statistics, and
Internet technology, its counterpart in evolutionary informatics remains
``scattered, poorly documented, and in formats that impede discovery and
integration'' (Parr et al. 2011). Our goal in developing the
\texttt{treebase} package is to provide steps to reduce these challenges
through programmatic and interactive access between the repositories
that store this data and the software tools commonly used to analyse
them.

The R statistical environment (R Development Core Team 2012) has become
a dominant platform for researchers using phylogenetic data to address a
rapidly expanding set of questions in ecological and evolutionary
processes. These methods include, but are not limited to, ancestral
state reconstruction (Paradis 2004; Butler and King 2004),
diversification analysis (Paradis 2004; Rabosky 2006; Harmon et al.
2008), identifying trait dependent speciation and extinction rates,
(Fitzjohn 2010; Goldberg, Lancaster, and Ree 2011; Stadler 2011b),
quantifying the rate and tempo of trait evolution (Butler and King 2004;
Harmon et al. 2008; Eastman et al. 2011), identifying evolutionary
influences and proxies for community ecology (Webb, Ackerly, and Kembel
2008; Kembel et al. 2010), connecting phylogeny data to climate patterns
(Warren, Glor, and Turelli 2008; Evans et al. 2009), and simulation of
speciation and character evolution (Harmon et al. 2008; Stadler 2011a;
Boettiger, Coop, and Ralph 2012), as well as various manipulations and
visualizations of phylogenetic data (Paradis 2004; Schliep 2010;
Jombart, Balloux, and Dray 2010; Revell et al. 2011). A more
comprehensive list of R packages by analysis type is available on the
phylogenetics taskview,
\href{http://cran.r-project.org/web/views/Phylogenetics.html}{http://cran.r-project.org/web/views/Phylogenetics.html}.
A few programs for applied phylogenetic methods are written for
environments outside the R environment, incuding Java (Maddison and
Maddison 2011), MATLAB (Blomberg, Garland, and Ives 2003) and Python
(Sukumaran and Holder 2010) and online interfaces (Martins 2004).

TreeBASE (\href{http://treebase.org}{http://treebase.org}) is an online
repository of phylogenetic data (e.g.~trees of species, populations, or
genes) that have been published in a peer-reviewed academic journal,
book, thesis or conference proceedings (Sanderson et al. 1994; Morell
1996). The database can be searched through an online interface which
allows users to find a phylogenetic tree from a particular publication,
author or taxa of interest. TreeBASE provides an application programming
interface (API) that lets computer applications make queries to the
database. Our \texttt{treebase} package uses this API to create a direct
link between this data and the R environment. This has several immediate
and important benefits:

\begin{enumerate}
\item \emph{Data discovery.} Users can leverage the rich, higher-level
  programming environment provided by the R environment to better
  identify data sets appropriate for their research by iteratively
  constructing queries for datasets that match appropriate metadata
  requirements.
\item \emph{Programmatic data access.} Many tasks that are theoretically
  made possible by the creation of the TreeBASE repository are not
  pursued because they would be too laborious for an exploratory
  analysis. The ability to use programmatic access across data sets to
  automatically download and perform a reproduciblye and systematic
  analysis using the rich set of tools available in R opens up new
  avenues for research.
\item \emph{Automatic updating}. The TreeBASE repository is expanding
  rapidly. The scriptable nature of analyses run with our
  \texttt{treebase} package means that a study can be rerun on the
  latest version of the repository without additional effort but with
  potential new information.
\end{enumerate}

\subsection{Programmatic Web Access}

The Treebase repository makes data accessible by Web queries through a
RESTful (REpresentational State Transfer) interface, which supplies
search conditions in the address URL. The repository returns the
requested data in XML (extensible markup language) format. The
\texttt{treebase} package uses the \texttt{RCurl} package (Lang 2012a)
to make queries over the Web to the repository, and the \texttt{XML}
package (Lang 2012b) to parse the Web page returned by the repository
into meaningful R data objects. While these querying and parsing
functions comprise most of the code provided in the \texttt{treebase}
package, they are hidden from the end user who can interact with these
rich data retrieval and manipulation tools to access data from these
remote repositories in much the same way as data locally available on
the users hard-disk.

\subsection{Basic queries}

The \texttt{treebase} package allows these queries to be made directly
from R, just as a user would make them from the Web browser. This
enables a user to construct more complicated filters than permitted by
the Web interface, and allows the user to maintain a record of the
queries they used to collect their data as an R script. Scripting the
data-gathering process helps reduce errors and assists in replicating
the analysis later, either by the authors or other researchers (Peng et
al. 2011).

The \texttt{search\_treebase} function forms the base of the
\texttt{treebase} package. Table 1 lists each of the types of queries
available through the \texttt{search\_treebase} function. This list can
also be found in the function documentation through the R command
\texttt{?search\_treebase}.\\Any of the queries available on the Web
interface can now be made directly from R, including downloading and
importing a phylogeny into the R interface. For instance, one can search
for phylogenies containing dolphin taxa, ``Delphinus,'' or all
phylogenies submitted by a given author, ``Huelsenbeck'' using the R
commands

\begin{Shaded}
\begin{Highlighting}[]
    \KeywordTok{search_treebase}\NormalTok{(}\StringTok{"Delphinus"}\NormalTok{, }\DataTypeTok{by=}\StringTok{"taxon"}\NormalTok{)}
    \KeywordTok{search_treebase}\NormalTok{(}\StringTok{"Huelsenbeck"}\NormalTok{, }\DataTypeTok{by=}\StringTok{"author"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}
This function returns the matching phylogenies into R as an R object,
ready for analysis. The package documentation provides many examples of
possible queries.

\ctable[caption = Queries available in \texttt{search\_treebase}. The
first argument is the keyword used in the query such as an author's name
and the second argument indicates the type of query (\emph{i.e.}
``author'')., pos = H, center, botcap]{ll}
{% notes
}
{% rows
\FL
search ``by='' & purpose
\ML
abstract & search terms in the publication abstract
\\\noalign{\medskip}
author & match authors in the publication
\\\noalign{\medskip}
subject & Matches in the subject terms
\\\noalign{\medskip}
doi & The unique object identifier for the publication
\\\noalign{\medskip}
ncbi & NCBI identifier number for the taxon
\\\noalign{\medskip}
kind.tree & Kind of tree (Gene tree, species tree, barcode tree)
\\\noalign{\medskip}
type.tree & Type of tree (Consensus or Single)
\\\noalign{\medskip}
ntax & Number of taxa in the matrix
\\\noalign{\medskip}
quality & A quality score for the tree, if it has been rated.
\\\noalign{\medskip}
study & Match words in the title of the study or publication
\\\noalign{\medskip}
taxon & Taxon scientific name
\\\noalign{\medskip}
id.study & TreeBASE study ID
\\\noalign{\medskip}
id.tree & TreeBASE's unique tree identifier (Tr.id)
\\\noalign{\medskip}
id.taxon & Taxon identifier number from TreeBase
\\\noalign{\medskip}
tree & The title for the tree
\LL
}

\subsection{Accessing all phylogenies}

For certain applications a user may wish to download all the available
phylogenies from TreeBASE. Using the \texttt{cache\_treebase} function
allows a user to download a local copy of all trees. Because direct
database dumps are not available from treebase.org, this function has
intentional delays to avoid overtaxing the TreeBASE servers, and should
be allowed a full day to run.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{treebase <- }\KeywordTok{cache_treebase}\NormalTok{()}
\end{Highlighting}
\end{Shaded}
Once run, the cache is saved compactly in memory where it can be easily
and quickly restored. For convenience, the \texttt{treebase} package
comes with a copy already cached, which can be loaded into memory.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{data}\NormalTok{(treebase)}
\end{Highlighting}
\end{Shaded}
All of the examples shown in this manuscript are run as shown using the
\texttt{knitr} package for authoring dynamic documents (Xie 2012), which
helps ensure the results shown are reproducible. These examples can be
updated by copying and pasting the code shown into the R terminal, or by
recompiling the entire manuscript from the source files found on the
development Web page for the TreeBASE package,
\href{https://github.com/ropensci/treebase}{github.com/ropensci/treebase}.
Data was accessed to produce the examples shown on Wed Jun 27 11:19:36
2012.

\section{Data discovery in TreeBASE}

Data discovery involves searching for existing data that meets certain
desired characteristics. Such searches take advantage of metadata --
summary information describing the data entries provided in the
repository. The Web repository uses separate interfaces (APIs) to access
metadata describing the publications associated with the data entered,
such as the publisher, year of publication, etc., and a different
interface to describe the metadata associated with an individual
phylogeny, such as the number of taxa or the kind of tree (\emph{e.g.}
Gene tree versus Species tree). The \texttt{treebase} package can query
these individual sources of metadata separately, but this information is
most powerful when used in concert -- allowing the construction of
complicated searches that cannot be automated through the Web interface.
The \texttt{metadata} function updates a list of all available metadata
from both APIs and returns this information as an R \texttt{data.frame}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{meta <- }\KeywordTok{metadata}\NormalTok{()}
\end{Highlighting}
\end{Shaded}
From the length of the metadata list we see that there are currently
3164 published studies in the database.

The fields provided by \texttt{metadata} are listed in Table II.

\ctable[caption = Columns of metadata available from the
\texttt{metadata} function, pos = H, center, botcap]{ll}
{% notes
}
{% rows
\FL
metadata field & description
\ML
Study.id & TreeBASE study ID
\\\noalign{\medskip}
Tree.id & TreeBASE's unique tree identifier
\\\noalign{\medskip}
kind & Kind of tree (Gene tree, species tree, barcode tree)
\\\noalign{\medskip}
type & Type of tree (Consensus or Single)
\\\noalign{\medskip}
quality & A quality score for the tree, if it has been rated.
\\\noalign{\medskip}
ntaxa & Number of taxa in the matrix
\\\noalign{\medskip}
date & Year the study was published
\\\noalign{\medskip}
author & First author in the publication
\\\noalign{\medskip}
title & The title of the publication
\LL
}

Metadata can also be used to reveal trends in the data deposition which
may be useful in identifying patterns or biases in research or emerging
potential types of data. As a simple example, we look at trends in the
submission patterns of publishers over time,

\begin{Shaded}
\begin{Highlighting}[]
    \NormalTok{date <- meta[[}\StringTok{"date"}\NormalTok{]] }
    \NormalTok{pub <- meta[[}\StringTok{"publisher"}\NormalTok{]]}
\end{Highlighting}
\end{Shaded}
Many journals have only a few submissions, so we will label any not in
the top ten contributing journals as ``Other'':

\begin{Shaded}
\begin{Highlighting}[]
    \NormalTok{topten <- }\KeywordTok{sort}\NormalTok{(}\KeywordTok{table}\NormalTok{(pub), }\DataTypeTok{decreasing=}\OtherTok{TRUE}\NormalTok{)[}\DecValTok{1}\NormalTok{:}\DecValTok{10}\NormalTok{]}
    \NormalTok{meta[[}\StringTok{"publisher"}\NormalTok{]][!(pub %in% }\KeywordTok{names}\NormalTok{(topten))] <- }\StringTok{"Other"}
\end{Highlighting}
\end{Shaded}
We plot the distribution of publication years for phylogenies deposited
in TreeBASE, color coding by publisher in Fig {[}fig:1{]}.

\begin{Shaded}
\begin{Highlighting}[]
  \KeywordTok{library}\NormalTok{(ggplot2) }
  \KeywordTok{ggplot}\NormalTok{(meta) + }\KeywordTok{geom_bar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(date, }\DataTypeTok{fill =} \NormalTok{publisher)) }
\end{Highlighting}
\end{Shaded}
\begin{figure}[htbp]
\centering
\includegraphics{../treebase/dates.pdf}
\caption{Histogram of publication dates by year, with the code required
to generate the figure.}
\end{figure}

Typically we are interested in the metadata describing the phylogenies
themselves rather than just in the publications in which they appeared.
Phylogenetic metadata includes features such as the number of taxa in
the tree, a quality score (if available), kind of tree (gene tree,
species tree, or barcode tree) or whether the phylogeny represents a
consensus tree from a distribution or just a single estimate.

Even simple queries can illustrate the advantage of interacting with
TreeBASE data through an R interface has over the Web interface. A Web
interface can only perform the tasks built in by design. For instance,
rather than performing six separate searches to determine the number of
consensus vs single phylogenies available for each king of tree, we can
construct a 2 by 2 table with a single line of code,

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{table}\NormalTok{(meta[[}\StringTok{"kind"}\NormalTok{]], meta[[}\StringTok{"type"}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}
\begin{table}[ht]
\begin{center}
\begin{tabular}{rrr}
  \hline
 & Consensus & Single \\ 
  \hline
Barcode Tree &   1 &   4 \\ 
  Gene Tree &  65 & 134 \\ 
  Species Tree & 2863 & 5857 \\ 
   \hline
\end{tabular}
\end{center}
\end{table}

\section{Reproducible computations}

Reproducible research has become a topic of increasing interest in
recent years, and facilitating access to data and using scripts that can
replicate analyses can help lower barriers to the replication of
statistical and computational results (Schwab, Karrenbach, and Claerbout
2000; Gentleman and Temple Lang 2004; Peng 2011). The \texttt{treebase}
package facilitates this process, as we illustrate in a simple example.

Consider the shifts in speciation rate identified by Derryberry et al.
(2011) on a phylogeny of ovenbirds and treecreepers. We will seek to not
only replicate the results the authors obtained by fitting the models
provided in the R package \texttt{laser} (Rabosky 2006), but also
compare them against methods presented in Stadler (2011b) and
implemented in the package \texttt{TreePar}, which permits speciation
models that were not available to Derryberry et al. (2011) at the time
of their study.

\subsection{Obtaining the tree}

By drawing on the rich data manipulation tools available in R which may
be familiar to the large R phylogenetics community, the
\texttt{treebase} package allows us to construct richer queries than are
possible through the TreeBASE Web interface alone.

The most expedient way to identify the data uses the digital object
identifer (doi) at the top of most articles, which we use in a call to
the \texttt{search\_treebase} function, such as

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{results <- }\KeywordTok{search_treebase}\NormalTok{(}\StringTok{"10.1111/j.1558-5646.2011.01374.x"}\NormalTok{, }\StringTok{"doi"}\NormalTok{) }
\end{Highlighting}
\end{Shaded}
The search returns a list, since some publications can contain many
trees. In this case our phylogeny is in the first element of the list.

Having imported the phylogenetic tree corresponding to this study, we
can quickly replicate their analysis of which diversification process
best fits the data. These steps can be easily implemented using the
phylogenetics packages we have just mentioned.

For instance, we can calculate the branching times of each node on the
phylogeny,

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{bt <- }\KeywordTok{branching.times}\NormalTok{(derryberry)}
\end{Highlighting}
\end{Shaded}
and then begin to fit each model the authors have tested, such as the
pure birth model,

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{yule = }\KeywordTok{pureBirth}\NormalTok{(bt)}
\end{Highlighting}
\end{Shaded}
or the birth-death model,

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{birth_death = }\KeywordTok{bd}\NormalTok{(bt)}
\end{Highlighting}
\end{Shaded}
The estimated models are now loaded into the active R session where we
can further explore them as we go along. The appendix shows the
estimation and comparison of all the models originally considered by
Derryberry et al. (2011).

In this fast-moving field, new methods often become available between
the time of submission and time of publication of a manuscript. For
instance, the more sophisticated models introduced in Stadler (2011b)
were not used in this study, but have since been made available in the
recent package, \texttt{TreePar}. These richer models permit a shift the
speciation or extinction rate to occur multiple times throughout the
course of the phylogeny.

We load the new method and format the phylogeny using the R commands:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(TreePar)}
\NormalTok{x <- }\KeywordTok{sort}\NormalTok{(}\KeywordTok{getx}\NormalTok{(derryberry), }\DataTypeTok{decreasing =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}
Here we consider models that have up to 4 different rates in Yule
models, (The syntax in \texttt{TreePar} is slightly cumbersome, the
{[}{[}2{]}{]} indicates where this command happens to store the output
models.)

As a comparison of speciation models is not the focus of this paper, the
complete code and explanation for these steps is provided as an
appendix. Happily, this analysis confirms the author's original
conclusions, even when the more general models of Stadler (2011b) are
considered.

\section{Analyses across many phylogenies}

Large scale comparative analyses that seek to characterize evolutionary
patterns across many phylogenies are increasingly common in phylogenetic
methods (\emph{e.g.} McPeek and Brown 2007; Phillimore and Price 2008;
McPeek 2008; Quental and Marshall 2010; Davies et al. 2011). Sometimes
referred to by their authors as meta-analyses, these approaches have
focused on re-analyzing phylogenetic trees collected from many different
earlier publications. This is a more direct approach than the
traditional concept of meta-analysis where statistical results from
earlier studies are weighted by their sample size without being able to
access the raw data. Because the identical analysis can be repeated on
the original data from each study, this approach avoids some of the
statistical challenges inherent in traditional meta-analyses summarizing
results across heterogeneous approaches.

To date, researchers have gone through heroic efforts simply to assemble
these data sets from the literature. As described in McPeek and Brown
(2007); (emphasis added)

\begin{quote}
One data set was based on 163 published species-level molecular
phylogenies of arthropods, chordates, and mollusks. A PDF format file of
each article was obtained, and a digital snapshot of the figure was
taken in Adobe Acrobat 7.0. This image was transferred to a PowerPoint
(Microsoft) file and printed on a laser printer. The phylogenies
included in this study are listed in the appendix. \emph{All branch
lengths were measured by hand from these printed sheets using dial
calipers.}

\end{quote}
Despite the recent emergence of digital tools that could now facilitate
this analysis without mechanical calipers, (\emph{e.g.} treesnatcher,
Laubach and von Haeseler 2007), it is easier and less error-prone to
pull properly formatted phylogenies from the database for this purpose.
Moreover, as the available data increases with subsequent publications,
updating earlier meta-analyses can become increasingly tedious. Using
\texttt{treebase}, a user can apply any analysis they have written for a
single phylogeny across the entire collection of suitable phylogenies in
TreeBASE, which can help overcome such barriers to discovery and
integration at this large scale. Using the functions we introduce
aboved, we provide a simple example that computes the gamma statistic of
Pybus and Harvey (2000), which provides an measure of when speciation
patterns differ from the popular birth-death model.

\subsection{Tests across many phylogenies}

A standard test of this is the gamma statistic of Pybus and Harvey
(2000) which tests the null hypothesis that the rates of speciation and
extinction are constant. The gamma statistic is normally distributed
about 0 for a pure birth or birth-death process, values larger than 0
indicate that internal nodes are closer to the tip then expected, while
values smaller than 0 indicate nodes farther from the tip then expected.
In this section, we collect all phylogenetic trees from TreeBASE and
select those with branch length data that we can time-calibrate using
tools available in R. We can then calculate the distribution of this
statistic for all available trees, and compare these results with those
from the analyses mentioned above.

The \texttt{treebase} package provides a compressed cache of the
phylogenies available in treebase. This cache can be automatically
updated with the \texttt{cache\_treebase} function,

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{treebase <- }\KeywordTok{cache_treebase}\NormalTok{()}
\end{Highlighting}
\end{Shaded}
which may require a day or so to complete, and will save a file in the
working directory named with treebase and the date obtained. For
convenience, we can load the cached copy distributed with the
\texttt{treebase} package:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{data}\NormalTok{(treebase)}
\end{Highlighting}
\end{Shaded}
We will only be able to use those phylogenies that include branch length
data. We drop those that do not from the data set,

\begin{Shaded}
\begin{Highlighting}[]
      \NormalTok{have <- }\KeywordTok{have_branchlength}\NormalTok{(treebase)}
      \NormalTok{branchlengths <- treebase[have]}
\end{Highlighting}
\end{Shaded}
Like most comparative methods, this analysis will require ultrametric
trees (branch lengths proportional to time, rather than to mutational
steps). As most of these phylogenies are calibrated with branch length
proportional to mutational step, we must time-calibrate each of them
first.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{timetree <- function(tree)}
    \KeywordTok{try}\NormalTok{( }\KeywordTok{chronoMPL}\NormalTok{(}\KeywordTok{multi2di}\NormalTok{(tree)) )}
\NormalTok{tt <- }\KeywordTok{drop_nontrees}\NormalTok{(}\KeywordTok{sapply}\NormalTok{(branchlengths, timetree))}
\end{Highlighting}
\end{Shaded}
At this point we have 1,396 time-calibrated phylogenies over which we
will apply the diversification rate analysis. Computing the gamma test
statistic to identify deviations from the constant-rates model takes a
single line,

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gammas <- }\KeywordTok{sapply}\NormalTok{(tt,  gammaStat)}
\end{Highlighting}
\end{Shaded}
and the resulting distribution of the statistic across available trees
is shown Fig 2. While researchers have often considered this statistic
for individual phylogenies, we are unaware of any study that has
visualized the empirical distribution of this statistic across thousands
of phylogenies. Both the overall distribution, which appears slightly
skewed towards positive values indicating increasing rate of speciation
near the tips, and the position and identity of outlier phylogenies are
patterns that may introduce new hypotheses and potential directions for
further exploration.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{qplot}\NormalTok{(gammas)+}\KeywordTok{xlab}\NormalTok{(}\StringTok{"gamma statistic"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}
\begin{figure}[htbp]
\centering
\includegraphics{../treebase/gammadist.pdf}
\caption{Distribution of the gamma statistic across phylogenies in
TreeBASE. Strongly positive values are indicative of an increasing rate
of evolution (excess of nodes near the tips), very negative values
indicate an early burst of diversification (an excess of nodes near the
root).}
\end{figure}

\section{Conclusion}

While we have focused on examples that require no additional data beyond
the phylogeny, a wide array of methods combine this data with
information about the traits, geography, or ecological community of the
taxa represented. In such cases we would need programmatic access to the
trait data as well as the phylogeny. The Dryad digital repository
(\href{http://datadryad.org}{http://datadryad.org}) is an effort in this
direction. While programmatic access to the repository is possible
through the \texttt{rdryad} package (Chamberlain, Boettiger, and Ram
2012), variation in data formatting must first be overcome before
similar direct access to the data is possible. Dedicated databases such
as FishBASE (\href{http://fishbase.org}{http://fishbase.org}) may be
another alternative, where morphological data can be queried for a list
of species using the \texttt{rfishbase} package (Boettiger). The
development of similar software for programmatic data access will
rapidly extend the space and scale of possible analyses.

The recent advent of mandatory data archiving in many of the major
journals publishing phylognetics-based research (\emph{e.g.} Fairbairn
2010; Piwowar, Vision, and Whitlock 2011; Whitlock et al. 2010), is a
particularly promising development that should continue to fuel the
trend of submissions seen in Fig. 1. Accompanied by faster and more
inexpensive techniques of NextGen sequencing, and the rapid expansion in
phylogenetic applications, we anticipate this rapid growth in available
phylogenies will continue. Faced with this flood of data, programmatic
access becomes not only increasingly powerful but an increasingly
necessary way to ensure we can still see the forest for all the trees.

\section{Acknowledgements}

CB wishes to thank S. Price for feedback on the manuscript, the TreeBASE
developer team for building and supporting the repository, and all
contributers to TreeBASE. CB is supported by a Computational Sciences
Graduate Fellowship from the Department of Energy under grant number
DE-FG02-97ER25308.


\input{../settings/boilerplate}


